### ISSUE TITLE
Consider adding `unrelated_type_checks` like with `unrelated_type_equality_checks`

### ISSUE THREAD

**Issue by ltOgt** [dart-lang/linter#5052]
The following is covered by `unrelated_type_equality_checks`:

```dart
void someFunc(int number, String string) {
  if (number == string) {} // linter gives a warning for unrelated types
}
```

But the following does not give a warning:
```dart
void someFunc(int number, String string) {
  if (number is String) {} // no warning for unrelated type check
}
```

<img width="1050" alt="Screenshot 2024-08-08 at 10 46 38" src="https://github.com/user-attachments/assets/ec96c588-f12f-4bf5-b6e0-d6d4f2887c43">

**Comment by srawlins** [comment]
My 2 cents: I think there enough occasions where you want to check if `a` is a type that is unrelated to it's static type, so I would not report the general case.

_But_ you can still report cases that will never be true, in the cases of final classes. The example above, "is this int a String?" is a perfect example.

Given the expression `x` has static type `T`, and you're checking `x is U` (or `x as U`, I believe):

* If `T` is final and `U` is neither a supertype of `T` nor a supertype of any of `T`'s subtypes (which can only be declared in `T`'s library), then the check will always be false.
* If `U` is final and `T` is neither a supertype of `U` nor a supertype of any of `U`'s subtypes (which can only be declared in `U`'s library), then the check will always be false. (I don't think I have this backwards...)

The same does not hold for _sealed_ classes, as their subclasses can be subclassed arbitrarily.

We can similarly report unnecessary checks, where `U` _is_ a supertype of `T`.

**Comment by FMorschel** [comment]
Inspired by this issue I created #5062 about `unrelated_type_casts`.

**Comment by FMorschel** [comment]
From @eernstg's [comment](https://github.com/dart-lang/linter/issues/5062#issuecomment-2438055010) on #5062:

> This issue is somewhat related: https://github.com/dart-lang/language/issues/4074
